apps/lockfile-explorer/src/commands/lint.ts

```diff
@@ -0,0 +1,167 @@
+import type { RushConfigurationProject } from '@microsoft/rush-lib/lib/api/RushConfigurationProject';
+import { FileSystem, JsonFile, JsonSchema } from '@rushstack/node-core-library';
+import type { CommandModule } from 'yargs';
+import { Colorize } from '@rushstack/terminal';
+import semver from 'semver';
+
+import lockfileLintSchema from '../schemas/lockfile-lint.schema.json';
+import { getShrinkwrapFileMajorVersion, parseDependencyPath } from '../utils/shrinkwrap';
+import { LockfileExplorerConfig } from '../constants/common';
+
+export interface ILintRule {
+  rule: 'restrict-versions';
+  project: string;
+  requiredVersions: Record<string, string>;
+}
+
+export interface ILockfileLint {
+  rules: ILintRule[];
+}
+
+async function checkVersionCompatibility(
+  shrinkwrapFileMajorVersion: number,
+  packages: Lockfile['packages'],
+  dependencyPath: string,
+  requiredVersions: Record<string, string>,
+  checkedDependencyPaths: Set<string>
+): Promise<void> {
+  if (packages && packages[dependencyPath] && !checkedDependencyPaths.has(dependencyPath)) {
+    checkedDependencyPaths.add(dependencyPath);
+    const { name, version } = parseDependencyPath(shrinkwrapFileMajorVersion, dependencyPath);
+    if (name in requiredVersions && !semver.satisfies(version, requiredVersions[name])) {
+      throw new Error(`ERROR: Detected inconsistent version numbers in package '${name}': '${version}'!`);
+    }
+
+    for (const [dependencyPackageName, dependencyPackageVersion] of Object.entries(
+      packages[dependencyPath].dependencies ?? {}
+    )) {
+      await checkVersionCompatibility(
+        shrinkwrapFileMajorVersion,
+        packages,
+        `/${dependencyPackageName}${shrinkwrapFileMajorVersion === 6 ? '@' : '/'}${dependencyPackageVersion}`,
+        requiredVersions,
+        checkedDependencyPaths
+      );
+    }
+  }
+}
+
+async function searchAndValidateDependencies(
+  rushConfiguration: RushConfiguration,
+  checkedProjects: Set<RushConfigurationProject>,
+  project: RushConfigurationProject,
+  requiredVersions: Record<string, string>
+): Promise<void> {
+  console.log(`Checking the project: ${project.packageName}.`);
+
+  const projectFolder: string | undefined = project.projectFolder;
+  const subspace: Subspace | undefined = project.subspace;
+  if (subspace && projectFolder) {
+    const shrinkwrapFilename: string = subspace.getCommittedShrinkwrapFilename();
+    const pnpmLockfileText: string = await FileSystem.readFileAsync(shrinkwrapFilename);
+    const doc = yaml.load(pnpmLockfileText) as Lockfile | LockfileV6;
+    const { importers, lockfileVersion, packages } = doc;
+    const shrinkwrapFileMajorVersion: number = getShrinkwrapFileMajorVersion(lockfileVersion);
+    const checkedDependencyPaths: Set<string> = new Set<string>();
+    for (const [relativePath, { dependencies }] of Object.entries(importers)) {
+      if (path.resolve(projectFolder, relativePath) === projectFolder) {
```

Comment:
Is it possible to make this comparison without using `path.resolve`? Functions on `path` are somewhat slow.

Reply:
I think it would be better to use `path.resolve` here, using other methods might be a bit hacky